### PR TITLE
Refactor DFE Sign In related code

### PR DESCRIPTION
**Pull request by scruti**
This work is part of an overhaul of how we structure/namespace the code related to external integrations before adding new integrations to the service.

This is a refactor, so the behaviour should stay the same. The principles behind the refactoring are:
- Define clear boundaries between usage of "lib" and "app/services" project folders:
  - The library folder is left to domain-agnostic code that could be extracted into a gem and doesn't depend on our service domain. In this case, the library folder is exclusively used for  DFE Sign In API-related code.
  - The services folder is used for code that depends on our domain. In this case, we use code relating to parsing, exporting and writing into DB the DSI user's info obtained from the API calls.

- Structure the DSI-related code following Object Oriented Programming principles.
  - Instead of having a collection of methods around, we grouped and extracted different DSI-related objects with different responsibilities.
  - Used modules/classes to namespace and group the DSI code.
  - Normalised the service objects public interface methods to `#call`.

## Before
image

## After
image
image

### CODE REVIEW

File: app/jobs/export_dsi_users_to_big_query_job.rb [hunk 1]
```diff
@@ -1,13 +1,10 @@
-require "export_dsi_users_to_big_query"
-require "export_dsi_approvers_to_big_query"
```
Comment (scruti) [inline app/jobs/export_dsi_users_to_big_query_job.rb:2]: This code is domain-dependant and has been moved from library to service objects. There is no need to require it anymore.

File: app/jobs/export_dsi_users_to_big_query_job.rb [hunk 2]
```diff
@@ -1,13 +1,10 @@
...
 class ExportDSIUsersToBigQueryJob < ApplicationJob
   queue_as :low
 
   def perform
     return if DisableExpensiveJobs.enabled?
 
-    ExportDSIUsersToBigQuery.new.run!
-    ExportDSIApproversToBigQuery.new.run!
+    Publishers::DfeSignIn::BigQueryExport::Users.new.call
+    Publishers::DfeSignIn::BigQueryExport::Approvers.new.call
```
Comment (scruti) [inline app/jobs/export_dsi_users_to_big_query_job.rb:8]: - Used modules/classes to structure and namespace the code. 
- Normalised the service objects public interface to `.call`

File: app/services/publishers/dfe_sign_in/big_query_export/base.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+require "dfe_sign_in/api"
+require "google/cloud/bigquery"
+
+module Publishers::DfeSignIn::BigQueryExport
+  class Base
+    include DfeSignIn::API
+    include Publishers::DfeSignIn::Parsing
```
Comment (scruti) [inline app/services/publishers/dfe_sign_in/big_query_export/base.rb:7]: Using composition through modules, we can combine library code (API calls) with the domain-related parsing module.

File: app/services/publishers/dfe_sign_in/update_users_in_db.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
+require "dfe_sign_in/api"
+
+module Publishers::DfeSignIn
+  class UpdateUsersInDb
+    include DfeSignIn::API
+    include Publishers::DfeSignIn::Parsing
```
Comment (scruti) [inline app/services/publishers/dfe_sign_in/update_users_in_db.rb:6]: This class has nothing to do with BigQ exports. 
Both share the DSI API and Parsing code through composition while keeping the BigQ export-related code exclusively within its namespace/classes.

File: lib/dfe_sign_in/api.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
+require "dfe_sign_in/api/request"
+require "dfe_sign_in/api/response"
+
+module DfeSignIn
+  module API
```
Comment (scruti) [inline lib/dfe_sign_in/api.rb:5]: The original API code has been completely restructured into different modules/classes.

File: lib/dfe_sign_in/api/response.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+      end
+
+      def number_of_pages
+        raise(error_message) if response["numberOfPages"].nil?
+
+        response["numberOfPages"]
+      end
+
+      def users
+        raise(DfeSignIn::API::Response::NilUsersError, error_message) if users_nil_or_empty?
```
Comment (scruti) [inline lib/dfe_sign_in/api/response.rb:19]: As this error was always raised when iterating over the response users on a higher abstraction level, I have moved it directly to the `#users` call to improve cohesion and reduce the coupling from the class calling for the results.

**Review by KyleMacPherson — APPROVED** [review]

**Pull request merged**